### ISSUE TITLE
Proposal to add a data source property to Agrume (provides data to collection view)

### ISSUE THREAD

**Issue by mcrakhman** [JanGorman/Agrume#15]
First of all, thanks for the great library! I really enjoy it. To fulfill requirements of my project I made certain small additions to the library which may be useful for the others, though I am not 100% sure of that. That's why I made an issue first, before filing a pull request :smiley: 

So my proposal is: to add a protocol, e.g. `AgrumeDataSource` (the property is set upon `init`) which provides particular data to collection view based on the indexes of currently visible cells. Thus the methods of `AgrumeDataSource` are called in the data source methods of `collectionView` (dequeue cell etc). 

```Swift
public protocol AgrumeDataSource : class {
	func getTotalImagesCount () -> Int
	func downloadImageForIndex (index: Int, completion: (UIImage) -> ())
}
```
Below see implementation of certain methods which will shed some light on usage of the protocol:
```Swift
public func collectionView (collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
	if let dataSource = agrumeDataSource {
		return dataSource.getTotalImagesCount ()
	} else {
		return images?.count > 0 ? images.count : imageURLs.count
	}
}
```

```Swift
public func collectionView (collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell
... 
	if let images = self.images {
		cell.image = images[indexPath.row]

	} else if let dataSource = agrumeDataSource {
		spinner.alpha = 1
	
		dataSource.downloadImageForIndex (indexPath.row) { [weak self] image in 
			for currentIndexPath in collectionView.indexPathsForVisibleItems() { 

	// we need to check whether the user still sees the relevant image 
	// otherwise there is no need to show the image 
	
				if currentIndexPath.row == indexPath.row {
					cell.image = image
					self?.spinner.alpha = 0
				}
			}
		}
}
...
```

Because my application constantly loaded thumbnails from dropbox I needed the data in the collection view to be reloaded from time to time (to update the amount of rows). I implemented `willDisplayCell` collection view delegate method which called `reloadData` when the user reached the last row.  

```Swift

public func collectionView (collectionView: UICollectionView, willDisplayCell cell: UICollectionViewCell, forItemAtIndexPath indexPath: NSIndexPath) {
		
	didScroll?(index: indexPath.row)
		
	if let dataSource = agrumeDataSource {
		let totalItems = collectionView.numberOfItemsInSection(0)
			
		if indexPath.row == totalItems - 1 
						 && totalItems != dataSource.getTotalImagesCount() {

	// we check whether more images were loaded by the data source
	// otherwise no need to reload

			dispatch_async (dispatch_get_main_queue()) { [weak self] in
				self?.collectionView.reloadData()
			}
			
		}
	}
}
```

If you think that these additions are acceptable and/or needed please tell, and I will make a pull request. And, of course I appreciate any comments on this :smile_cat: 

Thanks!

**Comment by JanGorman** [comment]
Thanks for taking the time to write and your suggestion.

If I understand you correctly, the idea is another initialiser that that doesn't take a fixed set of `image{s|Urls}` but rather passes in a data source, along the lines of:
```swift
public convenience init(dataSource: AgrumeDataSource, …) {}
```
And then queries the data source like you've outlined.

The main use case is to enable adding items to the collection view presented by Agrume. I'm open to that, sure.

Some suggestions about implementation/method naming:

`protocol AgrumeDataSource : class` Any particular reason why this is required to be a class?

Instead of `getTotalImagesCount` I'd prefer `numberOfItems`. The get prefix was always frowned upon in Objective-C, I'd keep it that way in Swift :) – also more closely matches the nomenclature used by UICollectionView. And then why not make it a variable instead of function, Swift protocols allow for that.

`downloadImageForIndex` shouldn't imply that the image is downloaded, rather something like just `imageForIndex`.

The whole looping over `collectionView.indexPathsForVisibleItems` could be simplified?

```swift
if collectionView.indexPathsForVisibleItems().contains(indexPath) {
   cell.image = image
   self?.spinner.hidden = true
}
```
I guess any other cell that's dequeued takes care of cleaning up the spinner hiding

Anyways, open feel free to open a PR and we can work on it from there. Thanks again.

**Comment by mcrakhman** [comment]
Thanks for your reply!

Please see my answers below.

I made this protocol a class because that way we can provide a weak reference to it. Thus we can be sure that the `AgrumeDataSource` lifecycle is not affected by `Agrume`. Though on a second thought maybe keeping strong reference is better, so `AgrumeDataSource` is kept alive as long as `Agrume` is (this is better from the user's standpoint). So agree with deleting a reference to class.

Your comments re naming convention is fine with me.

Agree with simplifying the cycle. Though I think its better to use `spinner.alpha = 0` than `spiner.hidden = true`. 

When I implemented the function with `hidden` (true/false) instead of `alpha` (0,1) the spinner indicator did not show up (neither `dispatch_async` helped). Moreover `alpha` property was used in the similar situation when the URL's were downloaded (so I just used the same approach).   

I will make a PR taking into account all that was discussed. Thanks!

**Comment by JanGorman** [comment]
Resolved with #16
